opencart 1.5.6.4/upload/admin/controller/module/smaily_for_opencart.php

```diff
@@ -189,6 +231,35 @@ public function index() {
         $this->response->setOutput($this->render());
     }
 
+    protected function handleLayoutSaving() {
+        if (!$this->user->hasPermission('modify', 'module/smaily_for_opencart')) {
+            return;
+        }
+        $optin_form_settings = [];
+        // Load Smaily admin model for saving settings.
+        $this->load->model('smailyforopencart/admin');
+        // Check if layout settings are in POST & that it's an array.
+        if (!empty($this->request->post['smaily_optin_form_settings']) && is_array($this->request->post['smaily_optin_form_settings'])) {
+            // Declare layout module settings in POST.
+            $optin_form_settings = $this->request->post['smaily_optin_form_settings'];
+            // Loop over each array and validate array keys.
+            foreach ($smaily_modules as &$module) {
+                // Check if layout_id is number, if not default to 0.
+                $module['layout_id'] = (isset($module['layout_id']) && is_numeric($module['layout_id'])) ? $module['layout_id'] : '0';
+                // Check if position is one out of 4 available, if not default to content_bottom
+                $module['position'] = (isset($module['position']) && in_array(
+                    $module['position'], array('content_top', 'content_bottom', 'column_left', 'column_right'), true
+                )) ? $module['position'] : 'content_bottom';
+                // Check if status is 1 or 0, if none default to 0.
+                $module['status'] = (isset($module['status']) && $module['status'] === '1' || $module['status'] === '0') ? $module['status'] : '0';
+                // Check if sort order is a number, if not default to 1.
+                $module['sort_order'] = (isset($module['sort_order']) && is_numeric($module['sort_order'])) ? $module['sort_order'] : '';
+            }
+        }
+        // Save layout settings.
+        $this->model_smailyforopencart_admin->editSettingValue('smaily', 'smaily_optin_form_settings', $optin_form_settings);
```

Comment:
You are saving OPT-IN form settings even if there is no posted data. Why?

Reply:
If you remove every layout setting you do not send a post. So if there is no post, I will clear the settings.